### ISSUE TITLE
getOriginal() does not return original null values when they changed

### ISSUE THREAD

**Issue by nchachereau** [cakephp/cakephp#6613]
**Expected behavior:** When using an entity's method `getOriginal()` (defined in \Cake\Datasource\EntityTrait), one would expect, as the API says, to get the original value of a property.

I think even some other parts of the code in EntityTrait expect `getOriginal()` to return `null` even if the value has changed. See especially these lines in `extractOriginalChanged()`:

            $original = $this->getOriginal($property);
            if ($original !== null && $original !== $this->get($property)) {
                $result[$property] = $original;
            }

**Observed behavior:** However, if the original value was `null`, and the property has changed, `getOriginal()` will return the new value instead of `null`.

**Comment by nchachereau** [comment]
The problem comes from using `isset()` which will return `false` whether the property is `null` or hasn't been defined.

Proposed fix: use `array_key_exists` instead:

In `getOriginal()`:

```diff
-        if (isset($this->_original[$property])) {
+        if (array_key_exists($property, $this->_original)) {
             return $this->_original[$property];
         }
```

In `set()`:

```diff
-            if (!isset($this->_original[$p]) &&
-                isset($this->_properties[$p]) &&
+            if (!array_key_exists($p, $this->_original) &&
+                array_key_exists($p, $this->_properties) &&
                 $this->_properties[$p] !== $value
             ) {
                 $this->_original[$p] = $this->_properties[$p];
             }
```

(Please excuse me for not attaching an actual patch or even a merge request... this is (for me) much faster.)

**Comment by AD7six** [comment]
> this is (for me) much faster.

I think this is a duplicate of other tickets or at least has come up before.

There aren't any test cases that appear to cover what you're describing, I suspect the proposed fix won't work _correctly_ in all cases owing to the way null is assumed to mean missing in other places, [such as this](https://github.com/cakephp/cakephp/pull/6472/files).

**Comment by nchachereau** [comment]
Thanks for this feedback. 

As for the duplicate, I searched for other tickets before reporting. I'm pretty sure this hasn't come up yet.

As for the problems the fix might provoke, it could be, but I thought about the problem before suggesting this. At the very least, the code you link to will NOT be affected: extractOriginal() and extractOriginalChanged() simply continue behaving as they do now (that is, not reporting original null values - which might also surprise some users, btw.) As I pointed out, the fact that they test for getOriginal() possibly returning null values is a strong indicator that this is the expected behavior.

And no, there are not any test cases on this yet. Which is partly why it was faster suggesting this patch instead of writing a formal PR.

**Comment by markstory** [comment]
I don't see this as a risky change personally. It makes sense that you would want to be able to get the original null values too.

**Comment by lorenzo** [comment]
I think it is a safe change as well

**Comment by markstory** [comment]
I'll have a pull request up tonight.

**Comment by markstory** [comment]
Fixed with #6626. Thanks for the diff @nchachereau

**Comment by nchachereau** [comment]
Thank *you* for fixing this!!
